Updates, bug fixes, tests, docs

LiamGoodacre opened the pull request:
- `pulp test` failed to run due to no `Test.Main`
  - renamed `Test.TemplateString` to `Test.Main`
- added test case for injection bug
- reimplemented substitution to account for injection
- added test case for malicious keys
- fix malicious keys via an object-as-a-map with exclamation mark key
  encoding
- added `templateBy` to allow transformation of keys
  - used by the 'Safe' template
- removed unnecessary module names from JavaScript
- added some examples to docs

Any thoughts?

LiamGoodacre committed:
Updates, bug fixes, tests, docs

- `pulp test` failed to run due to no `Test.Main`
  - added `Test.Main`
  - moved `Test.TemplateString` to `Test.Data.TemplateString`
- added test case for injection bug
- reimplemented substitution to account for injection
- added test case for malicious keys
- fix malicious keys via an object-as-a-map with exclamation mark key
  encoding
- added `templateBy` to allow transformation of keys
  - used by the 'Safe' template
- removed unnecessary module names from JavaScript
- added some examples to docs
- add myself to CONTRIBUTORS

Risto-Stevcev commented:
@LiamGoodacre Nice! this is awesome :)

The only thing I'm not sure about is renaming `Test.TemplateString` to `Test.Main`. I did that on purpose so that all modules have corresponding test modules in the same path.

For example: `src/Foo/Bar.purs` is module `Foo.Bar`, so it has a corresponding `test/Foo/Bar.purs` called `Test.Foo.Bar`). You can run the `Test.TemplateString` unit tests with `pulp test -m Test.TemplateString`.

What are you thoughts on that?

LiamGoodacre commented:
Ah I see.  Yes that makes sense.  Other libraries tend to have both: a test file per ~module, and a main file importing all the others.  How do you feel about having:

``` purescript
-- test/Data/TemplateString.purs
module Test.Data.TemplateString where
main = {- tests for Data.TemplateString -}
```

``` purescript
-- test/Main.purs
module Test.Main where
import Test.Data.TemplateString as TemplateString
main = TemplateString.main
```
- `test/Data/TemplateString.purs`
  - module `Test.Data.TemplateString` with tests for `Data.TemplateString`
- `test/Main.purs` - imports `Test.Data.TemplateString`

My argument here is that `pulp test` should just work 😄

Alternatively, if you don't like that, how about adding a `package.json` with a `test` script:

``` json
{
  "scripts": {
    "test": "pulp test -m Test.TemplateString"
  }
}
```

Such that `npm run test` works?

Or at the very least, documentation in the README on how to run the tests.

Risto-Stevcev commented:
Yeah I definitely think your solution to have `Test.Main` module run the `Test.Data.TemplateString` makes sense. Occasionally I forget to run `-m` and then it takes me a minute to figure out why the tests are failing.

LiamGoodacre commented:
Just added myself to `CONTRIBUTORS.md` too.  I added bullet points as the rendered output was formatted like this for some reason:

> Risto Stevcev
> Dario Oddenino Liam Goodacre (https://github.com/LiamGoodacre)

:S

Risto-Stevcev commented:
Ah, yeah you have to add two spaces at the end of Dario's line because it's markdown. No worries, I'll fix it real quick.

Thanks for PR!

LiamGoodacre commented:
Thanks!

The two spaces thing sounds a little fragile, some people have editors which strip spaces from the end of lines when they save!

Risto-Stevcev commented:
@LiamGoodacre Yeah I agree. You should check out my `github-markdown-live` library. You just run `github-markdown-live -f [filename].md` and then it creates a websocket server that renders your markdown in github markdown format, and updates live whenever you make a change :)
